public/API/API_GetGameProgression.php

```diff
@@ -0,0 +1,177 @@
+ *   int        TimesUsedInHardcoreUnlockMedian  number of unlocks analyzed for the MedianTimeToUnlockHardcore metric
+ *   int        MedianTimeToUnlock               median number of seconds required to unlock this achievement from starting to play the game
+ *   int        MedianTimeToUnlockHardcore       median number of seconds required to unlock this achievement in hardcore from starting to play the game
+ */
+
+use App\Models\Achievement;
+use App\Models\Game;
+use App\Models\PlayerAchievement;
+use App\Models\PlayerGame;
+use App\Models\PlayerProgressReset;
+use App\Models\PlayerSession;
+use App\Platform\Enums\PlayerProgressResetType;
+
+$gameId = (int) request()->query('i');
+$preferHardcore = (int) request()->query('h');
+
+$game = Game::with('system')->find($gameId);
+if (!$game) {
+    return response()->json([], 404);
+}
+
+// ===== basic game information =====
+$coreSet = $game->gameAchievementSets()->core()->first()?->achievementSet;
+
+$response = [
+    'ID' => $game->ID,
+    'Title' => $game->Title,
+    'ConsoleID' => $game->system->ID,
+    'ConsoleName' => $game->system->Name,
+    'ImageIcon' => $game->ImageIcon,
+    'NumDistinctPlayers' => $game->players_total,
+    'TimesUsedInBeatMedian' => $game->times_beaten,
+    'TimesUsedInHardcoreBeatMedian' => $game->times_beaten_hardcore,
+    'MedianTimeToBeat' => $game->median_time_to_beat,
+    'MedianTimeToBeatHardcore' => $game->median_time_to_beat_hardcore,
+    'TimesUsedInCompletionMedian' => $coreSet?->times_completed ?? 0,
+    'TimesUsedInMasteryMedian' => $coreSet?->times_completed_hardcore ?? 0,
+    'MedianTimeToComplete' => $coreSet?->median_time_to_complete,
+    'MedianTimeToMaster' => $coreSet?->median_time_to_complete_hardcore,
+    'NumAchievements' => $game->achievements_published,
+    'Achievements' => [],
+];
+
+$achievements = $game->achievements()->published()->get();
+$achievementIds = $achievements->pluck('ID');
+$unlock_times = [];
+$unlock_hardcore_times = [];
+foreach ($achievementIds as $achievementId) {
+    $unlock_times[$achievementId] = [];
+    $unlock_hardcore_times[$achievementId] = [];
+}
+
+// ===== process the 100 most recent players to earn at least half of the achievements in the set =====
+$recentPlayerIds = PlayerGame::query()
+    ->where('game_id', $game->ID)
+    ->where($preferHardcore ? 'achievements_unlocked_hardcore' : 'achievements_unlocked', '>=', $game->achievements_published / 2)
+    ->orderByDesc('last_unlock_at')
+    ->limit(100)
+    ->pluck('user_id');
+
+$resets = PlayerProgressReset::query()
+    ->where('type', PlayerProgressResetType::Game)
+    ->where('type_id', $game->ID)
+    ->whereIn('user_id', $recentPlayerIds)
+    ->pluck('created_at', 'user_id');
+
+foreach ($recentPlayerIds as $playerId) {
+    $unlocks = PlayerAchievement::query()
+        ->where('user_id', $playerId)
+        ->whereIn('achievement_id', $achievementIds)
+        ->whereNull('unlocker_id')
+        ->get();
```

Comment:
Performance: this executes separate queries for each player in the loop. For 100 players, that's potentially 200+ queries (`PlayerAchievement` + `PlayerSession` for each).

https://github.com/user-attachments/assets/80845c13-2045-44f8-b2f7-fab9d4dd3430

We should probably lift up loading all the unlocks & sessions in a couple of batch queries outside this loop.

Reply:
Performance: this endpoint takes ~2s on my machine to resolve. That may be dramatically improved by resolving the N+1 issues noted above. If it isn't, it may be worth considering if caching should be used on this endpoint just to mitigate someone aggressively slamming the same ID over and over.

Reply:
This is significantly improved from the previous iteration. Processing time for game ID 228 decreased from ~3s to ~380ms on my local machine.

While playing around with the latest code, I came up with another optimization which may be worth considering:
```php
foreach ($allSessions->get() as $session) {
    // Check if we have enough samples for meaningful medians.
    $totalSamples = 0;
    foreach ($unlock_times as $times) {
        $totalSamples += count($times);
    }
    // If we have enough samples, stop processing.
    if ($totalSamples > 1000) {
        break;
    }
```

Reply:
I wouldn't feel comfortable using any sort of hard-coded number here. There are already 9 sets with more than 500 achievements. 1000 captured unlocks might not provide data for many achievements.

Conversely, if we're querying the last 100 players, that's an average of 10 unlocks per player. Over 90% of sets have more than 10 achievements.